util/priority_queue.go

```diff
@@ -100,7 +134,176 @@ func (pq *PriorityQueue) Dequeue() Op {
 		return nil
 	}
 
-	op := heap.Pop(&pq.queue).(Op)
-	delete(pq.hit, op.Key())
-	return op
+	item := heap.Pop(&pq.queue).(*item)
+	delete(pq.hit, item.payload.Key())
+	return item.payload
+}
+
+// DelayedCall is a function that we're not going to run yet.
+type DelayedCall struct {
+	clock      clockwork.Clock
+	f          func()
+	elapsed    <-chan time.Time
+	cancelled  chan struct{}
+	terminated chan struct{}
+}
+
+// DelayCall runs 'f' after 'delay'.
+func DelayCall(clock clockwork.Clock, delay time.Duration, f func()) *DelayedCall {
+	call := DelayedCall{
+		clock: clock,
+		f:     f,
+	}
+	call.reset(delay)
+	return &call
+}
+
+func (d *DelayedCall) loop() {
+	defer close(d.terminated)
+	for {
+		select {
+		case <-d.elapsed:
+			d.f()
+		case <-d.cancelled:
+			return
+		}
+	}
+}
+
+// Cancel the delayed call.
+func (d *DelayedCall) Cancel() {
+	close(d.cancelled)
+	<-d.terminated
+}
+
+func (d *DelayedCall) reset(delay time.Duration) {
+	d.cancelled = make(chan struct{})
+	d.terminated = make(chan struct{})
+	d.elapsed = d.clock.After(delay)
+	go d.loop()
+}
+
+// Reset the delayed call to run after 'delay' (as measured from now).
+func (d *DelayedCall) Reset(delay time.Duration) {
+	d.Cancel()
+	d.reset(delay)
+}
+
+// ScheduledItem is an item in a queue of scheduled items.
+type ScheduledItem interface {
+	Key() string
+	// Scheduled returns the earliest possible time the time is available for
+	// dequeueing.
+	Scheduled() time.Time
+}
+
+// scheduledOp adapts a ScheduledItem to an Op
+type scheduledOp struct {
+	ScheduledItem
+}
+
+// Priority implements Op.
+func (op scheduledOp) Priority() int64 {
+	return -op.Scheduled().Unix()
+}
+
+// SchedulingQueue is like a priority queue, but the first item is the oldest
+// scheduled item.
+type SchedulingQueue struct {
```

Comment:
Thanks for doing this @jml.

I'm still a bit concerned with this implementation, and I'm struggling to explain why.  The use of a queue with a 100m buffer as if it was an unlimited queue isn't inspiring confidence, neither is the tests failing.  I'm not even sure you're allowed to modify a variable that your selecting on within the select...

The chunk store implementation is very sensitive to the PriorityQueue, and I think I should take back what I said previously (asking you to use the priority queue and the condition).  Sharing code it not the go way!  I had a stab at implementing an actor-based scheduling queue (see https://github.com/weaveworks/cortex/blob/tom-ruler/ruler/scheduling_queue.go) and it passes all your tests...  Let me know what you think of this one - having the heap owned and only ever modified on a single go routine makes reasoning about the locking trivial, which I like.

Sorry for sending you around in circles on this one, it was not my intention.  I'm bad at explaining what it is about this implementation I don't like - lets chat about it tomorrow.